README.md

```diff
@@ -566,33 +568,148 @@ There are two events that we were not able to decode, for varying reasons:
 
 _Note however, that we don't have a clean way to trap the data and log it. See [Logging unmatched events](#logging-unmatched-events) for an example of how one might log such unmatched events_
 
-### Handling introduction of new fields in JSON
-
-The below example demonstrates the addition of a `CartId` property in a newer version of `CreateCart`. It's worth noting that
-deserializing `CartV1.CreateCart` into `CartV2.CreateCart` requires `CartId` to be an optional property or the property will
-deserialize into `null` which is an invalid state for the `CartV2.CreateCart` record in F# (F# `type`s are assumed to never be `null`).
-
-```
+<a name="upconversion"></a>
+### Handling versioning of events in F# with FsCodec
+
+As a system evolves, the types used for events will inevitably undergo changes too. There are thorough guides such as
+[Versioning in an Event Sourced System by Greg Young](https://leanpub.com/esversioning); this will only scratch the surface,
+with some key F# snippets.
+
+High level rules:
+  1. The most important rule of all is that you never want to relinquish Total Matching, i.e. never add a `_` catch all case
+to a match expression.
+  2. The simplest way to add a new field in a backward compatible manner is by adding it as an `option` and then using
+     pattern matching to handle presence or absence of the value. 
+  3. Where it becomes impossible to use the serialization-time conversion mechanisms such as [`JsonIsomorphism`](#jsonisimorphism)
```

Comment:
I'm not sure how JsonIsomorphism fits into the versioning of events. Can you provide an example of it being used for that purpose?

Reply:
Hm, don't have one to hand - see the section https://github.com/jet/fscodec#typesafeenum-fallback-converters-using-jsonisomorphism though
in that case, it uses `string` as the proxy type, but that can equally be a record type. So you define a record to consume what's there, and provide conversions to/from the domain side type, with as much logic as you want.

Reply:
example: https://github.com/jet/propulsion/blob/059dc1b80cc6468ae5610f7847cfcdfdf2035cdc/src/Propulsion.DynamoStore/AppendsIndex.fs#L17

Reply:
Adding link...